src/lexer/tokenizerEngine.ts

```diff
@@ -0,0 +1,160 @@
+      // Advance current position by matched whitespace length
+      this.index += matches[0].length;
+      return matches[0];
+    } else {
+      return '';
+    }
+  }
+
+  private getNextToken(previousToken?: Token): Token | undefined {
+    return (
+      this.matchToken(TokenType.BLOCK_COMMENT) ||
+      this.matchToken(TokenType.LINE_COMMENT) ||
+      this.matchToken(TokenType.COMMA) ||
+      this.matchToken(TokenType.OPEN_PAREN) ||
+      this.matchToken(TokenType.CLOSE_PAREN) ||
+      this.matchToken(TokenType.QUOTED_IDENTIFIER) ||
+      this.matchToken(TokenType.NUMBER) ||
+      this.matchReservedWordToken(previousToken) ||
+      this.matchPlaceholderToken(TokenType.NAMED_PARAMETER) ||
+      this.matchPlaceholderToken(TokenType.QUOTED_PARAMETER) ||
+      this.matchPlaceholderToken(TokenType.INDEXED_PARAMETER) ||
+      this.matchPlaceholderToken(TokenType.POSITIONAL_PARAMETER) ||
+      this.matchToken(TokenType.VARIABLE) ||
+      this.matchToken(TokenType.STRING) ||
+      this.matchToken(TokenType.IDENTIFIER) ||
+      this.matchToken(TokenType.DELIMITER) ||
+      this.matchToken(TokenType.OPERATOR)
+    );
+  }
+
+  private matchPlaceholderToken(tokenType: TokenType): Token | undefined {
+    if (tokenType in this.REGEX_MAP) {
+      const token = this.matchToken(tokenType);
+      const tokenRule = this.REGEX_MAP[tokenType]!;
+      if (token) {
+        if (tokenRule.key) {
+          return { ...token, key: tokenRule.key(token.value) };
+        }
+        return token; // POSITIONAL_PARAMETER does not have a key transform function
+      }
+    }
+    return undefined;
+  }
+
+  // private getEscapedPlaceholderKey({ key, quoteChar }: { key: string; quoteChar: string }): string {
+  //   return key.replace(new RegExp(escapeRegExp('\\' + quoteChar), 'gu'), quoteChar);
+  // }
+
+  private matchReservedWordToken(previousToken?: Token): Token | undefined {
+    // A reserved word cannot be preceded by a '.'
+    // this makes it so in "mytable.from", "from" is not considered a reserved word
+    if (previousToken?.value === '.') {
+      return undefined;
+    }
+
+    // prioritised list of Reserved token types
+    return (
+      this.matchToken(TokenType.RESERVED_CASE_START) ||
+      this.matchToken(TokenType.RESERVED_CASE_END) ||
+      this.matchToken(TokenType.RESERVED_COMMAND) ||
+      this.matchToken(TokenType.RESERVED_BINARY_COMMAND) ||
+      this.matchToken(TokenType.RESERVED_DEPENDENT_CLAUSE) ||
+      this.matchToken(TokenType.RESERVED_JOIN) ||
+      this.matchToken(TokenType.RESERVED_KEYWORD) ||
+      this.matchToken(TokenType.RESERVED_LOGICAL_OPERATOR) ||
+      this.matchToken(TokenType.RESERVED_JOIN_CONDITION)
+    );
+  }
+
+  // Shorthand for `match` that looks up regex from REGEX_MAP
+  private matchToken(tokenType: TokenType): Token | undefined {
+    return this.match({
+      type: tokenType,
+      regex: this.REGEX_MAP[tokenType]!.regex,
+      transform: this.REGEX_MAP[tokenType]!.value,
```

Comment:
typescript doesn't provide proper type narrowing for variable keys:
https://stackoverflow.com/questions/64616994/typescript-type-narrowing-not-working-for-in-when-key-is-stored-in-a-variable/64618261#64618261

Reply:
This is not really the same problem. Here we don't know what the value of `tokenType` is.

The root of the issue is that `REGEX_MAP` is defined as `Partial<>`, so it might contain all TokenTypes or it might contain none.

Even I, when reading this code, can't really tell for sure.

I see two solutions:

- Replace the use of `Partial<>` with the actual list of keys that are allowed.
- Check for existence of a key, like we're doing in `matchPlaceholderToken`.